Smarty - {htxt} blocks should not be evaluated unless needed

totten opened the pull request:
Overview
----------------------------------------

Fixes a common class of PHP warning about `*.hlp` files.

<img width="965" alt="Screen Shot 2023-02-23 at 2 01 56 PM" src="https://user-images.githubusercontent.com/1336047/221040913-b141f4f7-7adf-4a80-9de3-6ee5b1c0dee0.png">

Before
----------------------------------------

The `{htxt}` block appears to act _like_ a conditional expression. If you look at `function.htxt.php`, it checks the value of `$id` and conditionally enables or disables the content.

But it is not an effective conditional. The nested content is evaluated before it calls `smarty_block_htxt()`.

Consequently, if you try to show help for field `foo`, then it _actually_ evaluates the help message for _everything in the `.hlp` file_ (even if it's some quirky field that's not relevant and that requires extra data).

After
----------------------------------------

The `{htxt}` is literally like an `{if}`.

You don't get warnings about irrelevant data.

totten committed:
Smarty - {htxt} blocks should not be evaluated unless needed

eileenmcnaughton commented:
@totten not so fast

CRM_Case_Form_EmailTest::testOpeningEmailForm
Smarty error: [in CRM/Contact/Form/Task/Help/Email/id-from_email.hlp line 3]: syntax error: mismatched tag {/if}. expected {/crmScope} (opened line 1). (Smarty_Compiler.class.php, line 2337)

/home/jenkins/bknix-dfl/build/build-1/web/sites/all/modules/civicrm/packages/Smarty/Smarty.class.php:1100
/home/jenkins/bknix-dfl/build/build-1/web/sites/all/modules/civicrm/packages/Smarty/Smarty.class.php:1860
/home/jenkins/bknix-dfl/build/build-1/web/sites/all/modules/civicrm/packages/Smarty/Smarty_Compiler.class.php:2255
/home/jenkins/bknix-dfl/build/build-1/web/sites/all/modules/civicrm/packages/Smarty/Smarty_Compiler.class.php:2337
/home/jenkins/bknix-dfl/build/build-1/web/sites/all/modules/civicrm/packages/Smarty/Smarty_Compiler.class.php:486
/home/jenkins/bknix-dfl/build/build-1/web/sites/all/modules/civicrm/packages/Smarty/Smarty_Compiler.class.php:306
/home/jenkins/bknix-dfl/build/build-1/web/sites/all/modules/civicrm/packages/Smarty/Smarty.class.php:1499
/home/jenkins/bknix-dfl/build/build-1/web/sites/all/modules/civicrm/packages/Smarty/Smarty.class.php:1432
/home/jenkins/bknix-dfl/build/build-1/web/sites/all/modules/civicrm/packages/Smarty/Smarty.class.php:1271
/home/jenkins/bknix-dfl/build/build-1/web/sites/all/modules/civicrm/CRM/Core/Smarty.php:193

eileenmcnaughton commented on CRM/Core/Smarty.php:
```diff
@@ -144,6 +144,7 @@ private function initialize() {
       $this->default_modifiers[] = 'escape:"htmlall"';
     }
     $this->load_filter('pre', 'resetExtScope');
+    $this->load_filter('pre', 'htxtFilter');
```
as an aside - from memory this is one thing I am still stuck on with smarty 3 - I think this way of loading filters changes & I haven't figured out what to do instead

totten replied:
To my eye, the contract for pre-filters looks pretty similar [v2](https://www.smarty.net/docsv2/en/plugins.prefilters.postfilters.tpl) and [v3](https://www.smarty.net/docs/en/advanced.features.prefilters.tpl). I suppose the name `loadFilter()` is preferred in v3, but it looks like `load_filter()` is still provided (via `SmartyBC`).

totten commented:
```
CRM_Case_Form_EmailTest::testOpeningEmailForm
Smarty error: [in CRM/Contact/Form/Task/Help/Email/id-from_email.hlp line 3]: syntax error: mismatched tag {/if}. expected {/crmScope} (opened line 1). (Smarty_Compiler.class.php, line 2337)
```

Right-o. It was going line-by-line and failed with certain extra bits at the start of line (eg `{foo}{htxt id=...`). Switched to `preg_replace_callback()`, and that should work better. (This test passes, anyway.)

Also corrected a typo in the new error message.

agileware-justin commented:
Reported regression, https://lab.civicrm.org/dev/core/-/issues/4303
